File: notebooksjava/_03URLify.java
public class _03URLify {

    // 解法1：双ポインタ法（後ろから操作）— O(n) time, O(1) extra space
    public String replaceSpaces(String chars, int trueLength) {
        char[] s = chars.toCharArray();
        int spaceCount = 0;
        for (int i = 0; i < trueLength; i++) {
            if (s[i] == ' ') spaceCount++;
        }

        int newIndex = trueLength + spaceCount * 2;

        for (int i = trueLength - 1; i >= 0; i--) {
            if (s[i] == ' ') {
                s[newIndex - 1] = '0';
                s[newIndex - 2] = '2';
                s[newIndex - 3] = '%';
                newIndex -= 3;
            } else {
                s[newIndex - 1] = s[i];
                newIndex--;
            }
        }
        return new String(s).trim();
    }

    // 解法2：Javaの組み込みメソッド — O(n) time, O(n) space
    public String replaceSpacesPythonic(String s, int trueLength) {
        return s.substring(0, trueLength).replace(" ", "%20");
    }

    public static void main(String[] args) {
        _03URLify sol = new _03URLify();

        // replaceSpaces（双ポインタ）
        // 末尾に spaceCount*2 分の余白が必要（Python 実装と同じ前提）
        assert sol.replaceSpaces("Mr John Smith    ", 13).equals("Mr%20John%20Smith") : "two spaces";
        assert sol.replaceSpaces("hello", 5).equals("hello")                          : "no spaces";
        assert sol.replaceSpaces("   ", 1).equals("%20")                              : "single space";

        // replaceSpacesPythonic
        assert sol.replaceSpacesPythonic("Mr John Smith    ", 13).equals("Mr%20John%20Smith");
        assert sol.replaceSpacesPythonic("hello", 5).equals("hello");
        assert sol.replaceSpacesPythonic("   ", 1).equals("%20");

        System.out.println("All test cases passed ✅");
    }
}

File: notebooksc++/03-urlify.cpp
#include <iostream>
#include <string>
#include <cassert>

using namespace std;

// Solution 1: Using two-pointer approach from back - O(n) time, O(1) extra space
string replaceSpaces(string chars, int trueLength) {
    int spaceCount = 0;
    for (int i = 0; i < trueLength; i++) {
        if (chars[i] == ' ') {
            spaceCount++;
        }
    }
    
    int newIndex = trueLength + spaceCount * 2;
    
    for (int i = trueLength - 1; i >= 0; i--) {
        if (chars[i] == ' ') {
            chars[newIndex - 1] = '0';
            chars[newIndex - 2] = '2';
            chars[newIndex - 3] = '%';
            newIndex -= 3;
        } else {
            chars[newIndex - 1] = chars[i];
            newIndex--;
        }
    }
    
    // Remove trailing spaces
    while (!chars.empty() && chars.back() == ' ') {
        chars.pop_back();
    }
    
    return chars;
}

// Solution 2: Using simple string replacement
string replaceSpacesPythonic(const string& s, int trueLength) {
    string result = s.substr(0, trueLength);
    string replaced = "";
    
    for (char c : result) {
        if (c == ' ') {
            replaced += "%20";
        } else {
            replaced += c;
        }
    }
    
    return replaced;
}

int main() {
    // Test replaceSpaces
    assert(replaceSpaces("Mr John Smith    ", 13) == "Mr%20John%20Smith");
    assert(replaceSpaces("a b  ", 3) == "a%20b");
    assert(replaceSpaces("abc", 3) == "abc");
    
    // Test replaceSpacesPythonic
    assert(replaceSpacesPythonic("Mr John Smith    ", 13) == "Mr%20John%20Smith");
    assert(replaceSpacesPythonic("a b  ", 3) == "a%20b");
    assert(replaceSpacesPythonic("abc", 3) == "abc");
    
    cout << "All test cases passed ✅" << endl;
    return 0;
}
